File: scripts/algorithms/S/Self Crossing/Self Crossing.cpp
class Solution {
public:
    bool isSelfCrossing(vector<int>& distance) {
        if (distance.size() <= 3) return false; //only can have intersection with more than 4 lines

        distance.insert(distance.begin(), 0); //for the edge case: line i intersect with line i-4 at (0, 0)
        for (int i = 3; i < distance.size(); i++) {
            //check line i-3
            if (distance[i - 2] <= distance[i] && distance[i - 1] <= distance[i - 3]) return true;

            //check line i-5
            if (i >= 5) {
                if (distance[i - 1] <= distance[i - 3] && distance[i - 1] >= distance[i - 3] - distance[i - 5] 
                    && distance[i - 2] >= distance[i - 4] && distance[i - 2] <= distance[i - 4] + distance[i])
                    return true;
            }
        }
        return false;
    }
};

File: scripts/algorithms/S/Self Crossing/Self Crossing.py
class Solution:
    def isSelfCrossing(self, x):
        n = len(x)
        if n < 4: return False
        for i in range(3, n):
            if x[i] >= x[i-2] and x[i-1] <= x[i-3]: return True
            if i >= 4 and x[i-1]==x[i-3] and x[i]+x[i-4]>=x[i-2]: return True
            if i >= 5 and 0<=x[i-2]-x[i-4]<=x[i] and 0<=x[i-3]-x[i-1]<=x[i-5]: return True
        return False
        
